### PR TITLE
index: add RevWalk variant that can be filtered by generation

### PR DESCRIPTION
**Pull request by yuja**
# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have added myself to the contributors in `CHANGELOG.md` (optional)

### CODE REVIEW

File: lib/src/index.rs [hunk 1]
```diff
@@ -2130,6 +2171,81 @@ mod tests {
...
+        let mut index = MutableIndex::full(3);
+        // 9 7
+        // | |
+        // 8 6
+        // |/|
+        // 4 5
+        // | |
+        // 2 3
+        // |/
+        // 1
```
Comment (martinvonz) [inline lib/src/index.rs:2185]: I'm curious about a case like this:

```
5
|\
| 4
| |
2 3
|/
1
```

Will 1 match when requesting 2 generations before 5 *and* when requesting 3 generations before 5? I think it should (and I think they will as implemented).

**Review by martinvonz — APPROVED** [review]
Regarding syntax, I remember that we have previously talked about syntax for this and said that it's a nice property of the current revset grammar is that non-symbol atoms (if that's the right term) can only appear as function arguments. If we want to preserve that property, we can't have an operator like `foo~5` (even if `~` wasn't already taken) with that `5` is the a decimal number instead of a revset itself. I'm open to reconsidering that, though.

File: lib/src/index.rs (hunk 1 above)
Comment (yuja) [reply lib/src/index.rs]: Yes, they're basically the same as `5--` and `5---` respectively. Adjusted the test to include these cases.

**Comment by yuja** [comment]
> Regarding syntax, I remember that we have previously talked about syntax for this and said that it's a nice property of the current revset grammar is that non-symbol atoms (if that's the right term) can only appear as function arguments. If we want to preserve that property, we can't have an operator like `foo~5` (even if `~` wasn't already taken) with that `5` is the a decimal number instead of a revset itself.

I think a symbol could be reinterpreted as decimal in that example, but I'm not gonna introduce integer or integer range syntax atm. Typical ancestor query would be like `expr---` (nth parents) or `expr | expr-` (self and parents). It might be useful if we had concise syntax for the latter, but I have no idea.

**Pull request merged**